include/ginkgo/core/solver/multigrid.hpp

```diff
@@ -0,0 +1,417 @@
+                    return size_type{0};
+                };
+            } else if (parameters_.mg_level.size() > 1) {
+                mg_level_index_ = [](const size_type level, const LinOp *) {
+                    return level;
+                };
+            }
+        } else {
+            mg_level_index_ = parameters_.mg_level_index;
+        }
+        if (!parameters_.solver_index) {
+            if (parameters_.coarsest_solver.size() >= 1) {
+                solver_index_ = [](const size_type, const LinOp *) {
+                    return size_type{0};
+                };
+            }
+        } else {
+            solver_index_ = parameters_.solver_index;
+        }
+        const auto mg_level_len = parameters_.mg_level.size();
+        if (mg_level_len == 0) {
+            GKO_NOT_SUPPORTED(this);
+        } else {
+            // each mg_level can not be nullptr
+            for (size_type i = 0; i < mg_level_len; i++) {
+                if (parameters_.mg_level.at(i) == nullptr) {
+                    GKO_NOT_SUPPORTED(this);
+                }
+            }
+        }
+        // verify pre-related parameters
+        this->verify_legal_length(true, parameters_.pre_smoother.size(),
+                                  mg_level_len);
+        // verify post-related parameters when post does not use pre
+        this->verify_legal_length(!parameters_.post_uses_pre,
+                                  parameters_.post_smoother.size(),
+                                  mg_level_len);
+        // verify mid-related parameters when mid does not use pre/post.
+        this->verify_legal_length(
+            parameters_.mid_case == multigrid_mid_uses::mid,
+            parameters_.mid_smoother.size(), mg_level_len);
+
+        cycle_ = parameters_.cycle;
+        if (system_matrix_->get_size()[0] != 0) {
+            // generate on the existed matrix
+            this->generate();
+        }
+    }
+    void verify_legal_length(bool checked, size_type len, size_type ref_len)
+    {
+        if (checked) {
+            // len = 0 uses default behaviour
+            // len = 1 uses the first one
+            // len > 1 : must contain the same len as ref(mg_level)
+            if (len > 1 && len != ref_len) {
+                GKO_NOT_SUPPORTED(this);
+            }
+        }
+    }
+
+private:
+    std::shared_ptr<const LinOp> system_matrix_{};
+    std::shared_ptr<const stop::CriterionFactory> stop_criterion_factory_{};
+    std::vector<std::shared_ptr<const gko::multigrid::MultigridLevel>>
+        mg_level_list_{};
+    std::vector<std::shared_ptr<LinOp>> pre_smoother_list_{};
+    std::vector<std::shared_ptr<LinOp>> mid_smoother_list_{};
+    std::vector<std::shared_ptr<LinOp>> post_smoother_list_{};
+    std::shared_ptr<LinOp> coarsest_solver_{};
+    std::function<size_type(const size_type, const LinOp *)> mg_level_index_;
+    std::function<size_type(const size_type, const LinOp *)> solver_index_;
+    multigrid_cycle cycle_;
```

Comment:
```suggestion
```
Similarly for the other variables repeated from `parameters_`, except where you really need them to be separate.